src/components/drawing/index.js

```diff
@@ -275,10 +281,20 @@ function singlePointStyle(d, sel, trace, markerScale, lineScale, marker, markerL
         sel.style('stroke-width', lineWidth + 'px');
 
         var markerGradient = marker.gradient;
-        var gradientColor = markerGradient && (d.mgc || markerGradient.color);
-        var gradientType = markerGradient && (d.mgt || markerGradient.type);
+
+        var gradientType = d.mgt;
+        if(gradientType) perPointGradient = true;
+        else gradientType = markerGradient && markerGradient.type;
+
         if(gradientType && gradientType !== 'none') {
-            sel.call(drawing.gradient, gd, gradientType, fillColor, gradientColor);
+            var gradientColor = d.mgc;
+            if(gradientColor) perPointGradient = true;
+            else gradientColor = markerGradient.color;
+
+            var gradientID = 'g' + gd._fullLayout._uid + '-' + trace.uid;
+            if(perPointGradient) gradientID += '-' + i;
```

Comment:
Now we're making either one gradient for the whole trace (if we can) or one per point (if any of the attributes varies per point) and it will update if the attributes change. This means that several traces (or several points in one trace with array attributes) will no longer be able to share a gradient def, but I don't think there are any cases where this can blow up.